monitor/inclusion_checker.go

```diff
@@ -0,0 +1,195 @@
+	Help: "Number of seconds since the oldest unincorporated certificate was submitted",
+})
+
+var inclusionErrors = promauto.NewCounterVec(prometheus.CounterOpts{
+	Name: "inclusion_checker_errors",
+	Help: "Number of errors encountered while attempting to check for certificate inclusion",
+}, []string{"type"})
+
+type InclusionOptions struct {
+	Interval       time.Duration
+	FetchBatchSize int64
+	MaxGetEntries  int64
+}
+
+type inclusionClient interface {
+	GetSTH(context.Context) (*ct.SignedTreeHead, error)
+	GetEntries(ctx context.Context, start, end int64) ([]ct.LogEntry, error)
+}
+
+type inclusionChecker struct {
+	logger           *log.Logger
+	client           inclusionClient
+	logURI           string
+	db               storage.Storage
+	signatureChecker *ct.SignatureVerifier
+	clk              clock.Clock
+	logID            int64
+	stopChan         chan bool
+
+	interval      time.Duration
+	batchSize     int64
+	maxGetEntries int64
+}
+
+func (ic *inclusionChecker) run() {
+	go func() {
+		ticker := time.NewTicker(ic.interval)
+		for {
+			select {
+			case <-ic.stopChan:
+				return
+			case <-ticker.C:
+				err := ic.checkInclusion()
+				if err != nil {
+					ic.logger.Printf("!!! Checking certificate inclusion failed: %s", err)
+				}
+			}
+		}
+	}()
+}
+
+func (ic *inclusionChecker) stop() {
+	ic.logger.Printf("Stopping %s inclusionChecker", ic.logURI)
+	ic.stopChan <- true
+}
+
+func (ic *inclusionChecker) checkInclusion() error {
+	current, err := ic.db.GetIndex(ic.logID)
+	if err != nil {
+		inclusionErrors.WithLabelValues("getIndex").Inc()
+		return fmt.Errorf("error getting current log index for %q: %s", ic.logURI, err)
+	}
+
+	certs, err := ic.db.GetUnseen(ic.logID)
+	if err != nil {
+		inclusionErrors.WithLabelValues("getUnseen").Inc()
+		return fmt.Errorf("error getting unseen certificates from %q: %s", ic.logURI, err)
+	}
+	if len(certs) == 0 {
+		// nothing to do, don't advance the index
+		return nil
+	}
+
+	sth, err := ic.client.GetSTH(context.Background())
+	if err != nil {
+		inclusionErrors.WithLabelValues("getSTH").Inc()
+		return fmt.Errorf("error getting STH from %q: %s", ic.logURI, err)
+	}
+	newHead, entries, err := ic.getEntries(current, int64(sth.TreeSize))
+	if err != nil {
+		inclusionErrors.WithLabelValues("getEntries").Inc()
+		return fmt.Errorf("error retrieving entries from %q: %s", ic.logURI, err)
+	}
+
+	err = ic.checkEntries(certs, entries)
+	if err != nil {
+		inclusionErrors.WithLabelValues("checkEntries").Inc()
+		return fmt.Errorf("error checking retrieved entries for %q: %s", ic.logURI, err)
+	}
+
+	err = ic.db.UpdateIndex(ic.logID, newHead)
+	if err != nil {
+		inclusionErrors.WithLabelValues("updateIndex").Inc()
+		return fmt.Errorf("error updating current index for %q: %s", ic.logURI, err)
+	}
+
+	return nil
+}
+
+func min(a, b int64) int64 {
+	if a < b {
+		return a
+	}
+	return b
+}
+
+func (ic *inclusionChecker) getEntries(start, end int64) (int64, []ct.LogEntry, error) {
```

Comment:
Can you write a unit test for this function? The pagination logic seems correct to me but I'd feel better if there was a unit test for that logic as well to ensure we're getting all of the indexes in the range in the correct sized buckets.

Reply:
There already is one, `TestGetEntries`, was there something specific missing from that which you think needs adding?

Reply:
Ugh sorry! Github had collapsed that diff and I think I missed it entirely. Taking a :mag: of that now. Thanks!

Reply:
Existing test is great, thanks :+1: